t/proto_template/.github/workflows/upload.yml

```diff
@@ -0,0 +1,71 @@
+name: 'upload'
+
+on:
+  push:
+    branches:
+      - 'main'
+  workflow_dispatch:
+
+permissions:
+  contents: 'read'
+  id-token: 'write'
+
+concurrency:
+  group: '${{ github.workflow }}-${{ github.head_ref || github.ref }}'
+  cancel-in-progress: true
+
+env:
+  GO_VERSION: 'REPLACE_GO_VERSION'
+  WIF_PROVIDER: 'REPLACE_WIF_PROVIDER'
+  WIF_SERVICE_ACCOUNT: 'REPLACE_WIF_SERVICE_ACCOUNT'
+  PROJECT_ID: 'REPLACE_PROJECT_ID'
+  GO_GAR_REPOSITORY_NAME: 'REPLACE_GO_GAR_REPOSITORY_NAME'
+  GO_GAR_REPOSITORY_LOCATION: 'REPLACE_GO_GAR_REPOSITORY_LOCATION'
+  GO_MODULE_PATH: 'REPLACE_GO_MODULE_PATH'
+
+jobs:
+  ci:
+    uses: './.github/workflows/ci.yml'
+    permissions:
+      contents: 'read'
+      id-token: 'write'
+
+  deploy_go:
+    needs:
+      - 'ci'
+    runs-on: 'ubuntu-latest'
+    steps:
+      - name: 'Checkout'
+        uses: 'actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9' # ratchet:actions/checkout@v3
+
+      - id: 'auth'
+        name: 'Authenticate to Google Cloud'
+        uses: 'google-github-actions/auth@ef5d53e30bbcd8d0836f4288f5e50ff3e086997d' # ratchet:google-github-actions/auth@v1
+        with:
+          workload_identity_provider: '${{ env.WIF_PROVIDER }}'
+          service_account: '${{ env.WIF_SERVICE_ACCOUNT }}'
+
+      - name: 'Setup gcloud'
+        uses: 'google-github-actions/setup-gcloud@d51b5346f85640ec2aa2fa057354d2b82c2fcbce' # ratchet:google-github-actions/setup-gcloud@v1
+        with:
+          install_components: 'package-go-module'
+
+      - name: 'Setup Go'
+        uses: 'actions/setup-go@fac708d6674e30b6ba41289acaab6d4b75aa0753' # ratchet:actions/setup-go@v4
+        with:
+          go-version: '${{ env.GO_VERSION }}'
+
+      - name: 'Upload to GAR (go)'
+        env:
+          # version must have format vX.Y.Z https://cloud.google.com/artifact-registry/docs/go/store-go#package_and_upload_your_module
+          VERSION: 'v0.0.0-${{ github.sha }}'
```

Comment:
As a follow up maybe, will we allow custom versioning or is there a reason why we dont?

Reply:
i can make the prefix a variable and default to this value. In general this pattern for the version works. This version value was the source of all the pain atleast for go.

Reply:
I feel like people will want to version their protos, can do in a follow up PR after some testing.